### PR TITLE
Update the /identity-providers/ API to manage user-defined federated authenticators.

### PR DESCRIPTION
**Pull request by Thisara-Welmilla**
Issue: 
- https://github.com/wso2/product-is/issues/21216

This PR introduces improvements to the following APIs for managing user-defined federated authenticators:
- /identity-providers: Updated the request body for the PUT request to accept both authenticator configuration models.
- /identity-providers/{identity-provider-id}/federated-authenticators: Updated the request body for the PUT request to accept both authenticator configuration models.
- /identity-providers/{identity-provider-id}/federated-authenticators/{federated-authenticator-id}: Updated the response and request bodies for both GET and PUT requests to accommodate both models.

Service layer PR:
- https://github.com/wso2/carbon-identity-framework/pull/6105

### CODE REVIEW

File: components/org.wso2.carbon.identity.api.server.idp/org.wso2.carbon.identity.api.server.idp.v1/src/gen/java/org/wso2/carbon/identity/api/server/idp/v1/model/Endpoint.java [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+import javax.validation.Valid;
+import javax.xml.bind.annotation.*;
+
+public class Endpoint  {
+  
+    private String uri;
+    private AuthenticationType authentication;
+
+    /**
+    **/
```
Comment (malithie) [inline components/org.wso2.carbon.identity.api.server.idp/org.wso2.carbon.identity.api.server.idp.v1/src/gen/java/org/wso2/carbon/identity/api/server/idp/v1/model/Endpoint.java:40]: empty comment ?

File: components/org.wso2.carbon.identity.api.server.idp/org.wso2.carbon.identity.api.server.idp.v1/src/main/java/org/wso2/carbon/identity/api/server/idp/v1/core/ServerIdpManagementService.java [hunk 2]
```diff
@@ -3081,13 +3151,42 @@ private FederatedAuthenticator createFederatedAuthenticator(String authenticator
...
+            AuthenticationType authenticationType = new AuthenticationType();
+            authenticationType.setType(AuthenticationType.TypeEnum.fromValue(endpointConfig.getEndpointConfig()
+                    .getAuthentication().getType().toString()));
+
+            Endpoint endpoint = new Endpoint();
+            endpoint.setUri(endpointConfig.getEndpointConfig().getUri());
+            authenticator.setEndpoint(endpoint);
+        } catch (ClassCastException e) {
+            throw new IdentityProviderManagementServerException("Error occurred while resolving endpoint " +
+                    "configuration of the authenticator.", e);
```
Comment (malithie) [inline components/org.wso2.carbon.identity.api.server.idp/org.wso2.carbon.identity.api.server.idp.v1/src/main/java/org/wso2/carbon/identity/api/server/idp/v1/core/ServerIdpManagementService.java:3186]: Can we specify the authenticator id  and name ?

File: components/org.wso2.carbon.identity.api.server.idp/org.wso2.carbon.identity.api.server.idp.v1/src/main/java/org/wso2/carbon/identity/api/server/idp/v1/core/ServerIdpManagementService.java [hunk 3]
```diff
@@ -1758,42 +1764,52 @@ private MetaOutboundConnector createMetaOutboundConnector(ProvisioningConnectorC
...
             String defaultAuthenticator = federatedAuthenticatorRequest.getDefaultAuthenticatorId();
             FederatedAuthenticatorConfig defaultAuthConfig = null;
             List<FederatedAuthenticatorConfig> fedAuthConfigs = new ArrayList<>();
             for (FederatedAuthenticator authenticator : federatedAuthenticators) {
-                FederatedAuthenticatorConfig authConfig = new FederatedAuthenticatorConfig();
-                authConfig.setName(base64URLDecode(authenticator.getAuthenticatorId()));
-                authConfig.setDisplayName(getDisplayNameOfAuthenticator(authConfig.getName()));
-                authConfig.setEnabled(authenticator.getIsEnabled());
-
+                String authenticatorName = getDecodedAuthName(authenticator.getAuthenticatorId());
```
Comment (malithie) [inline components/org.wso2.carbon.identity.api.server.idp/org.wso2.carbon.identity.api.server.idp.v1/src/main/java/org/wso2/carbon/identity/api/server/idp/v1/core/ServerIdpManagementService.java:1776]: Shall we rename as getDecodedAuthenticatorName

File: components/org.wso2.carbon.identity.api.server.idp/org.wso2.carbon.identity.api.server.idp.v1/src/main/java/org/wso2/carbon/identity/api/server/idp/v1/core/ServerIdpManagementService.java [hunk 4]
```diff
@@ -2839,34 +2856,89 @@ private ProvisioningConnectorConfig createProvisioningConnectorConfig(String out
...
+                  FederatedAuthenticatorPUTRequest authenticator) throws IdentityProviderManagementClientException {
 
+        String authenticatorName = getDecodedAuthName(federatedAuthenticatorId);
+        FederatedAuthenticatorConfig authConfig;
         String definedByType = null;
         if (authenticator.getDefinedBy() != null) {
             definedByType = authenticator.getDefinedBy().toString();
         }
-        authConfig.setDefinedByType(resolveDefinedByType(authenticatorName, definedByType, false));
+        definedByType = resolveDefinedByType(authenticatorName, definedByType, false).toString();
```
Comment (malithie) [inline components/org.wso2.carbon.identity.api.server.idp/org.wso2.carbon.identity.api.server.idp.v1/src/main/java/org/wso2/carbon/identity/api/server/idp/v1/core/ServerIdpManagementService.java:2867]: Better not have a boolean passed to indicate if this an authenticator create or update flow.
Either break it into two or pass the respective operation as an input

File: components/org.wso2.carbon.identity.api.server.idp/org.wso2.carbon.identity.api.server.idp.v1/src/main/java/org/wso2/carbon/identity/api/server/idp/v1/core/ServerIdpManagementService.java [hunk 5]
```diff
@@ -2839,34 +2856,89 @@ private ProvisioningConnectorConfig createProvisioningConnectorConfig(String out
      * @return Federated authenticator config of the specified ID.
      */
     private FederatedAuthenticatorConfig updateFederatedAuthenticatorConfig(String federatedAuthenticatorId,
```
Comment (malithie) [inline components/org.wso2.carbon.identity.api.server.idp/org.wso2.carbon.identity.api.server.idp.v1/src/main/java/org/wso2/carbon/identity/api/server/idp/v1/core/ServerIdpManagementService.java:2858]: Given that based on the definedByType either system federated or user defined authenticator configs can be built better separate that to two.

Or can use a factory to get different authenticator configs built based on the type

File: components/org.wso2.carbon.identity.api.server.idp/org.wso2.carbon.identity.api.server.idp.v1/src/main/java/org/wso2/carbon/identity/api/server/idp/v1/core/ServerIdpManagementService.java (hunk 3 above)
Comment (Thisara-Welmilla) [reply components/org.wso2.carbon.identity.api.server.idp/org.wso2.carbon.identity.api.server.idp.v1/src/main/java/org/wso2/carbon/identity/api/server/idp/v1/core/ServerIdpManagementService.java]: Addressed.

File: components/org.wso2.carbon.identity.api.server.idp/org.wso2.carbon.identity.api.server.idp.v1/src/main/java/org/wso2/carbon/identity/api/server/idp/v1/core/ServerIdpManagementService.java (hunk 4 above)
Comment (Thisara-Welmilla) [reply components/org.wso2.carbon.identity.api.server.idp/org.wso2.carbon.identity.api.server.idp.v1/src/main/java/org/wso2/carbon/identity/api/server/idp/v1/core/ServerIdpManagementService.java]: Addressed.

File: components/org.wso2.carbon.identity.api.server.idp/org.wso2.carbon.identity.api.server.idp.v1/src/main/java/org/wso2/carbon/identity/api/server/idp/v1/core/ServerIdpManagementService.java (hunk 5 above)
Comment (Thisara-Welmilla) [reply components/org.wso2.carbon.identity.api.server.idp/org.wso2.carbon.identity.api.server.idp.v1/src/main/java/org/wso2/carbon/identity/api/server/idp/v1/core/ServerIdpManagementService.java:2847]: Addressed

File: components/org.wso2.carbon.identity.api.server.idp/org.wso2.carbon.identity.api.server.idp.v1/src/main/java/org/wso2/carbon/identity/api/server/idp/v1/core/ServerIdpManagementService.java (hunk 2 above)
Comment (Thisara-Welmilla) [reply components/org.wso2.carbon.identity.api.server.idp/org.wso2.carbon.identity.api.server.idp.v1/src/main/java/org/wso2/carbon/identity/api/server/idp/v1/core/ServerIdpManagementService.java]: Addressed.

File: components/org.wso2.carbon.identity.api.server.idp/org.wso2.carbon.identity.api.server.idp.v1/src/main/java/org/wso2/carbon/identity/api/server/idp/v1/core/ServerIdpManagementService.java [hunk 6]
```diff
@@ -1758,42 +1765,40 @@ private MetaOutboundConnector createMetaOutboundConnector(ProvisioningConnectorC
...
-                authConfig.setDisplayName(getDisplayNameOfAuthenticator(authConfig.getName()));
-                authConfig.setEnabled(authenticator.getIsEnabled());
-
-                String definedByType = null;
-                if (authenticator.getDefinedBy() != null) {
-                    definedByType = authenticator.getDefinedBy().toString();
+                String authenticatorName = getDecodedAuthenticatorName(authenticator.getAuthenticatorId());
+                String definedByType;
+                if (isNewFederatedAuthenticator) {
+                    definedByType = resolveDefinedByTypeForCreateFederatedAuthenticator(
```
Comment (malithie) [inline components/org.wso2.carbon.identity.api.server.idp/org.wso2.carbon.identity.api.server.idp.v1/src/main/java/org/wso2/carbon/identity/api/server/idp/v1/core/ServerIdpManagementService.java:1780]: resolveDefinedByTypeForCreateFederatedAuthenticator -> resolveDefinedByTypeToCreateFederatedAuthenticator

File: components/org.wso2.carbon.identity.api.server.idp/org.wso2.carbon.identity.api.server.idp.v1/src/main/java/org/wso2/carbon/identity/api/server/idp/v1/core/ServerIdpManagementService.java [hunk 7]
```diff
@@ -1758,42 +1765,40 @@ private MetaOutboundConnector createMetaOutboundConnector(ProvisioningConnectorC
...
-                authConfig.setName(base64URLDecode(authenticator.getAuthenticatorId()));
-                authConfig.setDisplayName(getDisplayNameOfAuthenticator(authConfig.getName()));
-                authConfig.setEnabled(authenticator.getIsEnabled());
-
-                String definedByType = null;
-                if (authenticator.getDefinedBy() != null) {
-                    definedByType = authenticator.getDefinedBy().toString();
+                String authenticatorName = getDecodedAuthenticatorName(authenticator.getAuthenticatorId());
+                String definedByType;
+                if (isNewFederatedAuthenticator) {
```
[the remark this answers is not in the document]
Comment (malithie) [reply components/org.wso2.carbon.identity.api.server.idp/org.wso2.carbon.identity.api.server.idp.v1/src/main/java/org/wso2/carbon/identity/api/server/idp/v1/core/ServerIdpManagementService.java:1779]: However I feel having these two methods separated is also good and it's clear.

Check if we can get to top level

File: components/org.wso2.carbon.identity.api.server.idp/org.wso2.carbon.identity.api.server.idp.v1/src/main/java/org/wso2/carbon/identity/api/server/idp/v1/core/ServerIdpManagementService.java [hunk 8]
```diff
@@ -2839,46 +2845,40 @@ private ProvisioningConnectorConfig createProvisioningConnectorConfig(String out
...
-
-        String definedByType = null;
-        if (authenticator.getDefinedBy() != null) {
-            definedByType = authenticator.getDefinedBy().toString();
-        }
-        authConfig.setDefinedByType(resolveDefinedByType(authenticatorName, definedByType, false));
+                  FederatedAuthenticatorPUTRequest authenticator) throws IdentityProviderManagementClientException {
+
+        String authenticatorName = getDecodedAuthenticatorName(federatedAuthenticatorId);
+        String definedByType = resolveDefinedByTypeForUpdateFederatedAuthenticator(authenticatorName).toString();
```
Comment (malithie) [inline components/org.wso2.carbon.identity.api.server.idp/org.wso2.carbon.identity.api.server.idp.v1/src/main/java/org/wso2/carbon/identity/api/server/idp/v1/core/ServerIdpManagementService.java:2851]: resolveDefinedByTypeForUpdateFederatedAuthenticator -> resolveDefinedByTypeToUpdateFederatedAuthenticator

File: components/org.wso2.carbon.identity.api.server.idp/org.wso2.carbon.identity.api.server.idp.common/src/main/java/org/wso2/carbon/identity/api/server/idp/common/Constants.java [hunk 9]
```diff
@@ -146,6 +146,18 @@ public enum ErrorMessage {
         ERROR_CODE_ERROR_LISTING_TRUSTED_TOKEN_ISSUERS("60021",
                 "Unable to list existing trusted token issuers.",
                 "Server encountered an error while listing the trusted token issuers."),
+        ERROR_CODE_ENDPOINT_PROVIDED_FOR_SYSTEM_AUTH("60039", "No endpoint configuration is allowed " +
+                "for system defined authenticators.", "No endpoint configuration must be " +
+                "provided for the system defined federated authenticators %s."),
+        ERROR_CODE_PROPERTIES_PROVIDED_FOR_USER_AUTH("60040", "No properties are allowed for " +
+                "user defined authenticators.", "No properties must be provided for the user defined " +
+                "federated authenticators %s."),
```
Comment (malithie) [inline components/org.wso2.carbon.identity.api.server.idp/org.wso2.carbon.identity.api.server.idp.common/src/main/java/org/wso2/carbon/identity/api/server/idp/common/Constants.java:154]: Can we simplify the error message of these two cases. It's almost similar to a description

File: components/org.wso2.carbon.identity.api.server.idp/org.wso2.carbon.identity.api.server.idp.v1/src/main/java/org/wso2/carbon/identity/api/server/idp/v1/core/ServerIdpManagementService.java [hunk 10]
```diff
@@ -1758,42 +1766,41 @@ private MetaOutboundConnector createMetaOutboundConnector(ProvisioningConnectorC
...
-                                Constants.ErrorMessage.ERROR_CODE_INVALID_INPUT, " Duplicate properties are found in " +
-                                        "the request.");
-                    }
-                    List<Property> properties = authProperties.stream()
-                            .map(propertyToInternal)
-                            .collect(Collectors.toList());
-                    authConfig.setProperties(properties.toArray(new Property[0]));
+                if (DefinedByType.SYSTEM == definedByType) {
+                    validateAuthenticatorProperties(authenticatorName, authenticator.getProperties());
                 }
```
Comment (malithie) [inline components/org.wso2.carbon.identity.api.server.idp/org.wso2.carbon.identity.api.server.idp.v1/src/main/java/org/wso2/carbon/identity/api/server/idp/v1/core/ServerIdpManagementService.java:1789]: Can't we move this to the factory as that is what creates system and user defined authenticator configs ?

File: components/org.wso2.carbon.identity.api.server.idp/org.wso2.carbon.identity.api.server.idp.v1/src/main/java/org/wso2/carbon/identity/api/server/idp/v1/core/ServerIdpManagementService.java [hunk 11]
```diff
@@ -1758,42 +1766,41 @@ private MetaOutboundConnector createMetaOutboundConnector(ProvisioningConnectorC
...
-                if (authenticator.getDefinedBy() != null) {
-                    definedByType = authenticator.getDefinedBy().toString();
+                String authenticatorName = getDecodedAuthenticatorName(authenticator.getAuthenticatorId());
+                DefinedByType definedByType;
+                if (isNewFederatedAuthenticator) {
+                    definedByType = resolveDefinedByTypeToCreateFederatedAuthenticator(
+                            authenticator.getDefinedBy());
+                } else {
+                    definedByType = resolveDefinedByTypeToUpdateFederatedAuthenticator(authenticatorName);
                 }
```
Comment (malithie) [inline components/org.wso2.carbon.identity.api.server.idp/org.wso2.carbon.identity.api.server.idp.v1/src/main/java/org/wso2/carbon/identity/api/server/idp/v1/core/ServerIdpManagementService.java:1785]: We can extract this code for a function for clarity. Still keeping the two functions as they are usable as it is in other methods

File: components/org.wso2.carbon.identity.api.server.idp/org.wso2.carbon.identity.api.server.idp.v1/src/main/java/org/wso2/carbon/identity/api/server/idp/v1/core/ServerIdpManagementService.java [hunk 12]
```diff
@@ -2839,46 +2853,42 @@ private ProvisioningConnectorConfig createProvisioningConnectorConfig(String out
...
+                  FederatedAuthenticatorPUTRequest authenticator) throws IdentityProviderManagementClientException {
 
-        String definedByType = null;
-        if (authenticator.getDefinedBy() != null) {
-            definedByType = authenticator.getDefinedBy().toString();
+        String authenticatorName = getDecodedAuthenticatorName(federatedAuthenticatorId);
+        DefinedByType definedByType = resolveDefinedByTypeToUpdateFederatedAuthenticator(authenticatorName);
+        if (DefinedByType.SYSTEM == definedByType) {
+            validateAuthenticatorProperties(authenticatorName, authenticator.getProperties());
         }
```
Comment (malithie) [inline components/org.wso2.carbon.identity.api.server.idp/org.wso2.carbon.identity.api.server.idp.v1/src/main/java/org/wso2/carbon/identity/api/server/idp/v1/core/ServerIdpManagementService.java:2862]: Same comment above. Given the factor is there we should be able to move this to the factory

File: components/org.wso2.carbon.identity.api.server.idp/org.wso2.carbon.identity.api.server.idp.v1/src/main/java/org/wso2/carbon/identity/api/server/idp/v1/core/ServerIdpManagementService.java [hunk 13]
```diff
@@ -3081,13 +3089,42 @@ private FederatedAuthenticator createFederatedAuthenticator(String authenticator
...
+                        Arrays.stream(config.getProperties()).map(propertyToExternal).collect(Collectors.toList());
+                federatedAuthenticator.setProperties(properties);
+            } else {
+                federatedAuthenticator.setDefinedBy(FederatedAuthenticator.DefinedByEnum.USER);
+                resolveEndpointConfiguration(federatedAuthenticator, config);
+            }
+
         }
         return federatedAuthenticator;
     }
```
Comment (malithie) [inline components/org.wso2.carbon.identity.api.server.idp/org.wso2.carbon.identity.api.server.idp.v1/src/main/java/org/wso2/carbon/identity/api/server/idp/v1/core/ServerIdpManagementService.java:3105]: Doesn't this method now comes from the factory for system defined and user defined authenticators ?

File: components/org.wso2.carbon.identity.api.server.idp/org.wso2.carbon.identity.api.server.idp.v1/src/main/java/org/wso2/carbon/identity/api/server/idp/v1/core/ServerIdpManagementService.java (hunk 10 above)
Comment (Thisara-Welmilla) [reply components/org.wso2.carbon.identity.api.server.idp/org.wso2.carbon.identity.api.server.idp.v1/src/main/java/org/wso2/carbon/identity/api/server/idp/v1/core/ServerIdpManagementService.java:1789]: There are private methods in this class that are used by both validateAuthenticatorProperties and other methods. If we move them, we would either need to duplicate these private methods within the factory or relocate them to another utility class.

File: components/org.wso2.carbon.identity.api.server.idp/org.wso2.carbon.identity.api.server.idp.v1/src/main/java/org/wso2/carbon/identity/api/server/idp/v1/core/ServerIdpManagementService.java (hunk 13 above)
Comment (Thisara-Welmilla) [reply components/org.wso2.carbon.identity.api.server.idp/org.wso2.carbon.identity.api.server.idp.v1/src/main/java/org/wso2/carbon/identity/api/server/idp/v1/core/ServerIdpManagementService.java:3105]: In the factory, we are converting Auth API model -> Service layer fed model. Here what happens is Service layer fed model endpoint -> Auth API model enpoint

File: components/org.wso2.carbon.identity.api.server.idp/org.wso2.carbon.identity.api.server.idp.v1/src/main/java/org/wso2/carbon/identity/api/server/idp/v1/impl/FederatedAuthenticatorConfigBuilderFactory.java [hunk 14]
```diff
@@ -0,0 +1,433 @@
...
+                    String encodedData = null;
+                    int positionOfMetadataKey = -1;
+
+                    for (int i = 0; i < samlAuthenticatorProperties.size(); i++) {
+                        if (Constants.META_DATA_SAML.equals(samlAuthenticatorProperties.get(i).getName()) &&
+                                StringUtils.isNotBlank
+                                        (samlAuthenticatorProperties.get(i).getValue())) {
+                            validMetadataFound = true;
+                            encodedData = samlAuthenticatorProperties.get(i).getValue();
+                            positionOfMetadataKey = i;
```
Comment (malithie) [inline components/org.wso2.carbon.identity.api.server.idp/org.wso2.carbon.identity.api.server.idp.v1/src/main/java/org/wso2/carbon/identity/api/server/idp/v1/impl/FederatedAuthenticatorConfigBuilderFactory.java:267]: If metadata is found this loop can be broken right ?

**Review by malithie — APPROVED** [review]

**Review by malithie — APPROVED** [review]

**Pull request merged**